[core IR] Add a core decomposition for aten.all

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110042

## Context

Change the ref implementation of `aten.all` to only use other `torch` operators such that we can use it for the core ATen decomposition table. This will replace the decomposition for `aten.all` that was used specifically by Inductor.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

SS-JIA committed:
[core IR] Add a core decomposition for aten.all

[ghstack-poisoned]

SS-JIA committed:
Update on "[core IR] Add a core decomposition for aten.all"


## Context

Introduce a decomposition of `aten.all` into other `aten` operators that we can use for the core ATen decomposition table. This will replace the decomposition for `aten.all` that was used specifically by Inductor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

SS-JIA committed:
Update on "[core IR] Add a core decomposition for aten.all"


## Context

Introduce a decomposition of `aten.all` into other `aten` operators that we can use for the core ATen decomposition table. This will replace the decomposition for `aten.all` that was used specifically by Inductor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented on torch/_refs/__init__.py:
```diff
@@ -2224,7 +2224,6 @@ def _fn(*args, out=None, **kwargs):
 
-@register_decomposition(aten.all)
 @out_wrapper()
 def all(
```
Since this isn't calling a `prims.all`, we should update the reference instead of duplicating it as a decomposition. Decomposing into `any` makes a lot more sense than `sum` anyway, so this seems good.

SS-JIA replied:
I tried updating the refs to use the new decomposition, but it seems there is an issue with using `aten` ops in `_refs` implementations:

```
  File "/home/ssjia/Github/pytorch/torch/_refs/__init__.py", line 2235, in all
    result = aten.logical_not(aten.any(aten.logical_not(a)))
                                       ^^^^^^^^^^^^^^^^^^^
  File "/home/ssjia/Github/pytorch/torch/_ops.py", line 742, in __call__
    return self._op(*args, **kwargs or {})
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ssjia/Github/pytorch/torch/_prims/context.py", line 141, in __torch_function__
    raise RuntimeError(
RuntimeError: no _refs support for aten.logical_not
```

For decompositions included in the core ATen decomp table, we want to exclusively decompose to other ATen ops, i.e. avoid decomposing to prim ops, hence why ATen ops are referenced explicitly in the decomposition.

In this case, do you think it would be preferable to keep the ref implementation and the core decomposition separate? Or do you think it might be better to remove the ref implementation in favor of the core implementation?

peterbell10 replied:
You should be able to call `torch.logical_not` and `torch.any`, or even the methods.

SS-JIA replied:
thanks for the tip! let me try that...

SS-JIA replied:
Updated to change the ref implementation instead of introducing a new decomp (y)

SS-JIA committed:
Update on "[core IR] Add a core decomposition for aten.all"


## Context

Introduce a decomposition of `aten.all` into other `aten` operators that we can use for the core ATen decomposition table. This will replace the decomposition for `aten.all` that was used specifically by Inductor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

SS-JIA committed:
Update on "[core IR] Add a core decomposition for aten.all"


## Context

Introduce a decomposition of `aten.all` into other `aten` operators that we can use for the core ATen decomposition table. This will replace the decomposition for `aten.all` that was used specifically by Inductor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

SS-JIA committed:
Update on "[core IR] Add a core decomposition for aten.all"


## Context

Change the ref implementation of `aten.all` to only use other `torch` operators such that we can use it for the core ATen decomposition table. This will replace the decomposition for `aten.all` that was used specifically by Inductor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

SS-JIA committed:
Update on "[core IR] Add a core decomposition for aten.all"


## Context

Change the ref implementation of `aten.all` to only use other `torch` operators such that we can use it for the core ATen decomposition table. This will replace the decomposition for `aten.all` that was used specifically by Inductor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

SS-JIA commented:
Having some issues updating this PR with `ghstack`; closing in favor of https://github.com/pytorch/pytorch/pull/110083
